### PR TITLE
feat: Implement Segment sampling for expected errors

### PR DESCRIPTION
**Pull request by smilingkylan**
## **Description**

The purpose of this task is to reduce the instances of events being logged to Segment to only 1% of the current volume. The solution is to randomly select 1% of USERS (NOT EVENTS) to be tracked, and ignore the other 99%.

## **Related issues**

Fixes: [#4119](https://github.com/MetaMask/MetaMask-planning/issues/4119)

## **Manual testing steps**

To manually test you can opt-in to being tracked via the account creation user flow, then check MixPanel to see if the events were logged or not.

### **Before**

https://github.com/user-attachments/assets/63bfb33e-f2a1-4656-a60c-be1e633beff2

### **After**

https://github.com/user-attachments/assets/9c9d0f15-7927-429a-8143-65ec40779936

## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/core/Analytics/MetaMetrics.test.ts [hunk 1]
```diff
@@ -417,31 +439,30 @@ describe('MetaMetrics', () => {
...
+      await StorageWrapper.setItem(METAMETRICS_ID, VALID_UUID)
       const metaMetrics = TestMetaMetrics.getInstance();
       expect(await metaMetrics.configure()).toBeTruthy();
 
-      // reset the call count to checj no new calls are made
-      mockGet.mockClear();
+      // reset the call count on StorageWrapper.getItem
+      // we just want to know if the next `getMetaMetricsId` call
+      // will call StorageWrapper.getItem or not
+      StorageWrapper.getItem.mockClear(); // ?
```
Comment (smilingkylan) [inline app/core/Analytics/MetaMetrics.test.ts:457]: Typescript doesn't like when I use this method. I get `Property 'mockClear' does not exist on type '(key: string) => Promise<any>'.ts(2339)` but clearly the method exists

File: app/core/Analytics/MetaMetrics.ts [hunk 2]
```diff
@@ -785,6 +799,20 @@ class MetaMetrics implements IMetaMetrics {
...
+   */
+  getShouldTrackExpectedErrors = () => this.shouldTrackExpectedErrors;
+
+  /**
+   * Set whether to track expected "errors"
+   * @param shouldTrack - boolean
+   */
+  setShouldTrackExpectedErrors = (shouldTrack: boolean) => {
+    this.shouldTrackExpectedErrors = shouldTrack;
+  };
```
Comment (NicolasMassart) [inline app/core/Analytics/MetaMetrics.ts:815]: Q: does this have to be in MetaMetrics class? It could be in a util with trackErrorAsAnalytics, right?

File: app/core/Analytics/MetaMetrics.types.ts
Comment (NicolasMassart) [inline app/core/Analytics/MetaMetrics.types.ts:1]: Q: same as above, does this have to be in MetaMetrics class? It could be in a util with trackErrorAsAnalytics, right?

File: app/util/metrics/TrackError/trackErrorAsAnalytics.ts [hunk 3]
```diff
@@ -16,6 +16,8 @@ const trackErrorAsAnalytics = (
   errorMessage: string,
   otherInfo?: string,
 ) => {
+  if (!MetaMetrics.getInstance().getShouldTrackExpectedErrors()) return;
```
Comment (NicolasMassart) [inline app/util/metrics/TrackError/trackErrorAsAnalytics.ts:19]: C: the implementation is ok, but I don't think getShouldTrackExpectedErrors should be in MetaMetrics

File: ios/Podfile.lock [hunk 4]
```diff
@@ -1339,7 +1339,7 @@ SPEC CHECKSUMS:
   OpenSSL-Universal: ebc357f1e6bc71fa463ccb2fe676756aff50e88c
   Permission-BluetoothPeripheral: 34ab829f159c6cf400c57bac05f5ba1b0af7a86e
   PromisesObjC: f5707f49cb48b9636751c5b2e7d227e43fba9f47
-  RCT-Folly: 8dc08ca5a393b48b1c523ab6220dfdcc0fe000ad
+  RCT-Folly: 424b8c9a7a0b9ab2886ffe9c3b041ef628fd4fb1
```
Comment (NicolasMassart) [inline ios/Podfile.lock:1342]: Is this change related to this PR?

File: app/util/metrics/TrackError/trackErrorAsAnalytics.test.ts [hunk 5]
```diff
@@ -11,9 +10,12 @@ InteractionManager.runAfterInteractions = jest.fn(async (callback) =>
 
 const mockMetrics = {
   trackEvent: jest.fn(),
+  getShouldTrackExpectedErrors: jest.fn(() => true),
 };
 
-(MetaMetrics.getInstance as jest.Mock).mockReturnValue(mockMetrics);
+jest.mock('../../../core/Analytics/MetaMetrics', () => ({
+  getInstance: () => mockMetrics,
+}));
```
Comment (NicolasMassart) [inline app/util/metrics/TrackError/trackErrorAsAnalytics.test.ts:18]: C: yes, that's better, thanks!

**Review by NicolasMassart — DISMISSED** [review]
somme comments on the actual scope but the most significant part of my changes request is because of refactoring that should not be in this PR in my opinion. It's interesting refactoring but it's not needed for this PR specifically.

File: app/core/Analytics/MetaMetrics.ts (hunk 2 above)
Comment (smilingkylan) [reply app/core/Analytics/MetaMetrics.ts:815]: I figure the question here is whether or not we want to keep re-hashing (literally) every time we decide to use `trackErrorAsAnalytics`. If you prefer it as a utility function then I can do that, just confirm for me.

File: app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+describe('shouldTrackExpectedErrors', () => {
+  // Create a mock for the MetaMetrics instance with properly typed mock function
+  const mockMetaMetricsInstance = {
+    getMetaMetricsId: jest.fn<Promise<string | null>, []>(),
+  } as unknown as IMetaMetrics;
+
+  // Explicitly cast the mock function to Jest Mock type
+  const getMetaMetricsIdMock = mockMetaMetricsInstance.getMetaMetricsId as jest.Mock;
+
+  it('should return true when random number is less than threshold', async () => {
```
Comment (NicolasMassart) [inline app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts:31]: C: Follow tests naming guidelines
```suggestion
  it('returns true when error should be tracked', async () => {
```

File: app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    // Mock the return value for this specific test
+    (generateDeterministicRandomNumber as jest.Mock).mockReturnValue(0.006915);
+
+    const result = await shouldTrackExpectedErrors(mockMetaMetricsInstance);
+
+    expect(generateDeterministicRandomNumber).toHaveBeenCalledWith(TEST_IDS.UNDER_ONE_PERCENT);
+    expect(result).toBe(true);
+  });
+
+  it('should return false when random number is greater than threshold', async () => {
```
Comment (NicolasMassart) [inline app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts:46]:
```suggestion
  it('returns false when error should not be tracked', async () => {
```

File: app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    // Mock the return value for this specific test
+    (generateDeterministicRandomNumber as jest.Mock).mockReturnValue(0.975965);
+
+    const result = await shouldTrackExpectedErrors(mockMetaMetricsInstance);
+
+    expect(generateDeterministicRandomNumber).toHaveBeenCalledWith(TEST_IDS.OVER_NINETY_SEVEN_PERCENT);
+    expect(result).toBe(false);
+  });
+
+  it('should return false when threshold is 0', async () => {
```
Comment (NicolasMassart) [inline app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts:60]: S: Is this a edge case that has to be tested?
It seems to me it's already tested on the previous test. There's no specific code in the function saying "if zero", it's only a lower than test. So Zero in included and is not a special value.
I suggest deleting this test.

File: app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    // Mock the return value for this specific test
+    (generateDeterministicRandomNumber as jest.Mock).mockReturnValue(0.006915);
+
+    const result = await shouldTrackExpectedErrors(mockMetaMetricsInstance);
+
+    expect(generateDeterministicRandomNumber).toHaveBeenCalledWith(TEST_IDS.UNDER_ONE_PERCENT);
+    expect(result).toBe(false);
+  });
+
+  it('should return true when threshold is 1', async () => {
```
Comment (NicolasMassart) [inline app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts:73]: S: same, is 1 a specific value in the numbers range given the less than operator? whatever is over threshold is ok, 1 or more. It's just random num will never return more than one but do we care?

File: app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    // Mock the return value for this specific test
+    (generateDeterministicRandomNumber as jest.Mock).mockReturnValue(0.006915);
+
+    const result = await shouldTrackExpectedErrors(mockMetaMetricsInstance);
+
+    expect(generateDeterministicRandomNumber).toHaveBeenCalledWith(TEST_IDS.UNDER_ONE_PERCENT);
+    expect(result).toBe(true);
+  });
+
+  it('should use empty string when metaMetricsId is null', async () => {
```
Comment (NicolasMassart) [inline app/util/metrics/shouldTrackExpectedErrors/shouldTrackExpectedErrors.test.ts:86]:
```suggestion
  it('falls back using empty string metaMetricsId', async () => {
```

File: app/util/metrics/TrackError/trackErrorAsAnalytics.test.ts [hunk 11]
```diff
@@ -1,19 +1,29 @@
...
-InteractionManager.runAfterInteractions = jest.fn(async (callback) =>
-  callback(),
-);
+InteractionManager.runAfterInteractions = jest.fn(async (callback) => {
+  await callback();
+});
 
 const mockMetrics = {
   trackEvent: jest.fn(),
+  getMetaMetricsId: jest.fn(() => VALID_UUID),
```
Comment (NicolasMassart) [inline app/util/metrics/TrackError/trackErrorAsAnalytics.test.ts:21]: S: `getMetaMetricsId` must not be mocked here as it's not called directly from `trackErrorAsAnalytics` but only by `shouldTrackExpectedErrors`. The mock here is probably not even used as `shouldTrackExpectedErrors` is mocked and then should not even call `getMetaMetricsId`.
Suggestion: delete the mock for `getMetaMetricsId` here.

File: app/util/metrics/TrackError/trackErrorAsAnalytics.test.ts [hunk 12]
```diff
@@ -77,4 +90,13 @@ describe('trackErrorAsAnalytics', () => {
       .build();
     expect(mockMetrics.trackEvent).toHaveBeenCalledWith(expectedEvent);
   });
+
+  it('does not call trackEvent if shouldTrackExpectedErrors is false', async () => {
+    shouldTrackMock.shouldTrackExpectedErrors.mockResolvedValue(false);
```
Comment (NicolasMassart) [inline app/util/metrics/TrackError/trackErrorAsAnalytics.test.ts:95]: S: This mock should also be in all other tests in this suite but with value true. You can set it to true by default in a beforeEach if prefered.

**Review by NicolasMassart — COMMENTED** [review]
Some necessary changes on test cases and mocks, but the operational code looks good.

File: app/util/metrics/TrackError/trackErrorAsAnalytics.ts [hunk 13]
```diff
@@ -11,11 +12,15 @@ import { InteractionManager } from 'react-native';
  * @param errorMessage the error message for the original event
  * @param otherInfo other info to be logged
  */
-const trackErrorAsAnalytics = (
+const trackErrorAsAnalytics = async (
```
[the remark this answers is not in the document]
Comment (smilingkylan) [reply app/util/metrics/TrackError/trackErrorAsAnalytics.ts:15]: @NicolasMassart Are you seeing an eslint error in your IDE?

[the remark this answers is not in the document]
Comment (NicolasMassart) [reply app/util/metrics/TrackError/trackErrorAsAnalytics.ts:15]: yes I do and my IDE uses our project rules
image

[the remark this answers is not in the document]
Comment (NicolasMassart) [reply app/util/metrics/TrackError/trackErrorAsAnalytics.ts:15]: And for the trackErrorAsAnalytics when it's called in other files:
image
_Example here in app/components/Views/Login/index.js_

[the remark this answers is not in the document]
Comment (smilingkylan) [reply app/util/metrics/TrackError/trackErrorAsAnalytics.ts:15]: @Cal-L Can I get your opinion on this? The CI pipeline linting doesn't seem to have an issue with it but Nico has brought up the point that it could be a concern.

[the remark this answers is not in the document]
Comment (Cal-L) [reply app/util/metrics/TrackError/trackErrorAsAnalytics.ts:15]: CI should've caught this if it was a lint issue. @smilingkylan do you see the error when you run `yarn lint`? @NicolasMassart is that a lint warning or error? If it does appear as a warning, let's address it to reduce the noise

[the remark this answers is not in the document]
Comment (smilingkylan) [reply app/util/metrics/TrackError/trackErrorAsAnalytics.ts:15]: @Daniel-Cross Are you seeing linting issues or no?

[the remark this answers is not in the document]
Comment (Daniel-Cross) [reply app/util/metrics/TrackError/trackErrorAsAnalytics.ts:15]: Not seeing any issues in my linter and it's passing all the CI pipeline checks, which to me says it's okay to merge. Could just be a possible issue of an over zealous linter in Nico's case?

**Review by Daniel-Cross — APPROVED** [review]
